src/genie/libs/parser/iosxr/show_dampening.py

```diff
@@ -0,0 +1,241 @@
+"""
+
+# Python
+import re
+# Metaparser
+from genie.metaparser import MetaParser
+from genie.metaparser.util.schemaengine import Schema, Any, Optional
+
+# parser utils
+from genie.libs.parser.utils.common import Common
+
+# =============================================
+# Parser for 'show im dampening'
+# =============================================
+
+class ShowImDampeningSchema(MetaParser):
+    schema = {
+                'interface': {
+                    Any(): {
+                    'protocol': {
+                        Any(): {
+                        'capsulation': str,
+                        'penalty': int,
+                        'suppressed': str,
+                        }
+                    }
+                }
+            }
+        }
+                    
+class ShowImDampening(ShowImDampeningSchema):
+    """ Parser for show im dampening"""
+    
+    cli_command = 'show im dampening'
+
+    def cli(self,output=None):
+        """parsing mechanism: cli
+        Function cli() defines the cli type output parsing mechanism which
+        typically contains 3 steps: exe
+        cuting, transforming, returning
+        """
+        if output is None:
+            out = self.device.execute(self.cli_command)
+        else:
+            out = output
+            
+        result_dict = {}
+        result_dict.setdefault('interface', {})
+        
+        # GigabitEthernet0/0/0/1                                               2389 YES
+        # POS0/2/0/0                  <base>             ppp                      0 NO 
+        p1 = re.compile(r'^(?P<interface>.{1,26}) +(?P<prot>.{1,20}) +(?P<cap>.{1,10}) +(?P<pen>\d+) +(?P<sup>YES|NO)')
+
+        for line in out.splitlines():
+            if line:
+                line = line.strip()
+            else:
+                continue
+            
+            m = p1.match(line)
+            
+            if m:
+                group = m.groupdict()
+                interface=group['interface'].rstrip()
+                
+                # set "not_present" on any empty fields
+                for empty in group.keys():
+                    if group.get(empty).startswith(' '):
+                        group[empty]="not_present"
+                protocol=group['prot'].rstrip()
+                
+                if result_dict['interface'].get(interface)==None:
+                    result_dict['interface'][interface]={}
+                    result_dict['interface'][interface]['protocol']={}
+                result_dict['interface'][interface]['protocol'][protocol]={}
+                result_dict['interface'][interface]['protocol'][protocol].update({'capsulation': group['cap']})  
+                result_dict['interface'][interface]['protocol'][protocol].update({'penalty': int(group['pen'])})
+                result_dict['interface'][interface]['protocol'][protocol].update({'suppressed': group['sup'].lower()})                                     
```

Comment:
We can improve this like below as setdefault() will insert data if it doesnt exist but it also returns a reference to a point within a dictionary:
``` python
protocol_dict = result_dict.setdefault('interface', {}).setdefault(interface, {}).setdefault('protocol', {})
protocol_dict.update({'capsulation': group['cap']})
protocol_dict.update({'penalty': int(group['pen'])})
protocol_dict.update({'suppressed': group['sup'].lower()})

Reply:
Hi Daniel, That works good. I didn't know from this that protocol_dict would also update result_dict as well. They look to be two separate dictionaries but clearly they are not. I am guessing this is because they share the same object in memory? However, whilst this has solved one problem it has created another in that empty fields e.g. capsulation has a value of "None"  which is busting my schema since it's not a string. I could just convert it to a string or change the schema to any() but is that okay? The same has happened in the protocol field as well where any empty fields are present. Data, below: {'interface': {'GigabitEthernet0/0/0/0': {'protocol': {None: {'capsulation': None, 'penalty': 629, 'suppressed': 'NO'}}}, 'GigabitEthernet0/0/0/1': {'protocol': {None: {'capsulation': None, 'penalty': 2389, 'suppressed': 'YES'}}}, 'POS0/2/0/0': {'protocol': {None: {'capsulation': None, 'penalty': 0, 'suppressed': 'NO'}, '<base>': {'capsulation': 'ppp', 'penalty': 0, 'suppressed': 'NO'}, 'ipv4': {'capsulation': 'ipcp', 'penalty': 0, 'suppressed': 'NO'}}}}}

Reply:
For now, I've made the schema for capsulation to be Any() and it works fine. If this is not right then please let me know. Cheers.

Reply:
Hi @gwoodwa1 for any optional items in the regex you can do:
``` python
if group['cap']:
  protocol_dict.update({'capsulation': group['cap']})
```
As we dont want to insert anything if it never existed